Palette visualizer

gallottino opened the pull request:
Closes #89 

I also added a feature  _"debug"_ to put random values in the `palette_ram`. In can be useful to use this features also to some hard coded test for the render process.

To run clementine in debug mode you just need to run the command:
`just debug *rom_path/rom.gba*`

gallottino commented on emu/src/ppu.rs:
```diff
@@ -1,2 +1,133 @@
+    }
+
+    pub fn color_from_two_u8(&self, upper_byte: u8, lower_byte: u8) -> PaletteColor {
+        let mut upper: u32 = upper_byte.into();
```
I have some doubts about this: is it better to implement `Bits` trait also for u8?

gallottino commented:
Work in progress stuff:
- [ ] add documentation in `ppu.rs`
- [ ] add tests in `ppu.rs`

guerinoni commented on emu/src/ppu.rs:
```diff
@@ -1,2 +1,133 @@
 pub const DISPLAY_WIDTH: usize = 240;
 pub const DISPLAY_HEIGHT: usize = 160;
+pub const PALETTE_SIZE: usize = 16;
+
+pub const BG_PALETTE_ADDRESS: u32 = 0x05000000;
+pub const OBJ_PALETTE_ADDRESS: u32 = 0x05000200;
```
Maybe better declare those after `use...`

guerinoni commented on emu/src/ppu.rs:
```diff
@@ -1,2 +1,133 @@
+
+impl PixelProcessUnit {
+    pub fn new(internal_memory: Rc<RefCell<InternalMemory>>) -> Self {
+        Self { internal_memory }
```
In other iteration I would like to see this referring only to the part of video and not to entire `InternalMemory`, not important now, maybe we can just spit "logical memory" from "video memory"

gallottino replied:
I agree :)

guerinoni commented on ui/src/palette_visualizer.rs:
```diff
@@ -0,0 +1,132 @@
+use emu::{arm7tdmi::Arm7tdmi, gba::Gba};
+
+pub struct PaletteVisualizer {
+    gba: Arc<Mutex<Gba<Arm7tdmi>>>,
```
Maybe this should be `Arc<Mutex<Gba<Cpu>>>` or similar

gallottino replied:
Yes, but maybe `PaletteVisualizer` is a only Arm7tdmi tool. In other case we need to put the `InternalMemory` outside cpu or we need to implement in the trait a method for get it.

guerinoni commented on ui/src/palette_visualizer.rs:
```diff
@@ -0,0 +1,132 @@
+impl View for PaletteVisualizer {
+    fn ui(&mut self, ui: &mut egui::Ui) {
+        if let Ok(gba) = self.gba.lock() {
+            self.palettes = gba.cpu.ppu.get_palettes(self.palette_type.clone());
```
I think we can pass `&self.palette_type`

guerinoni reviewed (commented):
First round of review... There are some commits with "more" things in same context but... :P

gallottino commented on emu/src/render/palette_color.rs:
```diff
@@ -0,0 +1,128 @@
+    }
+}
+
+impl From<u16> for PaletteColor {
+    fn from(color_u16: u16) -> Self {
+        let color_u32: u32 = color_u16.into();
```
Not sure too. The problem is that u32 is not castable to PaletteColor as the color need to have only 15 bits and also we implemented the `Bits` trait only for u32.

AlessioC31 replied:
Can we implement `Bits` for `u16` too? I think it makes sense.

AlessioC31 commented on emu/src/render/palette_color.rs:
```diff
@@ -0,0 +1,129 @@
+pub enum PaletteType {
+    BG,
+    OBJ,
+    FULL,
```
What to we use `FULL` for?

AlessioC31 replied:
Ok, I think it is because the documentation says:

> Each BG and OBJ palette RAM may be either split into 16 palettes with 16 colors each, or may be used as a single palette with 256 colors.

So `FULL` is to access both areas as a single palette with 512 colors right? But wouldn't it be better to split this behavior into `BG_FULL` and `OBJ_FULL`?

Correct me if I'm wrong! Thanks!

gallottino replied:
Ah! yeah you are right, I missunderstood the docs.

AlessioC31 commented on emu/src/ppu.rs:
```diff
@@ -1,2 +1,130 @@
+            ],
+            PaletteType::FULL => match index {
+                0..=255 => self.get_array_color(index * 2, PaletteType::BG),
+                256..=551 => self.get_array_color((index - 256) * 2, PaletteType::OBJ),
```
Why `551`? We have 512 colors in total, right?

gallottino replied:
Yes, my bad :)

gallottino commented:
> Can we implement Bits for u16 too? I think it makes sense. 

@AlessioC31 
I don't know why I can't answer to your comment about implementing Bits traits for u16.

Anyway I also was thinking about it but we have the same methods  both for u32 and u16 and it seems to me a bit redundant. We can think a solution, but I prefer open an issue for this and iterate later. But I totally agree with you that it is a weird cast.

guerinoni commented:
> > Can we implement Bits for u16 too? I think it makes sense.
> 
> @AlessioC31 I don't know why I can't answer to your comment about implementing Bits traits for u16.
> 
> Anyway I also was thinking about it but we have the same methods both for u32 and u16 and it seems to me a bit redundant. We can think a solution, but I prefer open an issue for this and iterate later. But I totally agree with you that it is a weird cast.
> 
> P.s. Just a flash thought: we can force that the type for the `Bits` trait must implement `From<u16>` or something like this. But I would like to don't repeat the same code for all the u* types.

You can implement both and from u32 implementation call u16 function, in this way the core of logic it will stay only in one place :)

gallottino commented:
@guerinoni @AlessioC31 
I fixed some stuff. 

1. Now `Bits` trait can be used with all the uint types. I did a generical abstraction for all the cases.
2. I removed `FULL`  palette type at all
3. `From<u16> for PaletteColor` not doesn't cast anymore to u32
4. Minor stuff written in the previous comments

AlessioC31 reviewed (commented):
Why do we pass &PaletteType instead of PaletteType in some functions?

gallottino commented:
> Why do we pass &PaletteType instead of PaletteType in some functions?

To avoid `clone()` when you pass it many times in other functions. In other case, without clone the variable exits from the scope and it be lost. With `&PaletteType` we don't copy uselessly the value, but we are just passing its reference. For this specific case the difference is not so big because we don't need to pass by value a big amount of data, but anyway I think, when it is possible, this approach is better.

AlessioC31 commented:
> > Why do we pass &PaletteType instead of PaletteType in some functions?
> 
> To avoid `clone()` when you pass it many times in other functions. In other case, without clone the variable exits from the scope and it will be lost. With `&PaletteType` we don't copy uselessly the value, but we are just passing its reference. For this specific case the difference is not so big because we don't need to pass by value a big amount of data, but anyway I think, when it is possible, this approach is better.

But in this case `PaletteType` is practically an `int` (or something similar I suppose). I don't know how it works underneath in rust but if copying a `PaletteType` means copying an int and passing it by ref means passing a pointer, then passing it by value would be faster (you don't have the dereferencing part and since it's an `int` it will probably be passed via registers).

EDIT:
I checked with godbolt and it seems that in rust the code for the two ways are basically the same (with opt-level=0) because even when you pass by value it still dereferences (so it doesn't work as in C :D).

Also, in the case of optimizations they compile to the same assembly.

So I don't know, I would go for a `PaletteType` since it's "cleaner" but boh, no strong opinions since it's basically the same.

gallottino commented:
> > > Why do we pass &PaletteType instead of PaletteType in some functions?
> > 
> > 
> > To avoid `clone()` when you pass it many times in other functions. In other case, without clone the variable exits from the scope and it will be lost. With `&PaletteType` we don't copy uselessly the value, but we are just passing its reference. For this specific case the difference is not so big because we don't need to pass by value a big amount of data, but anyway I think, when it is possible, this approach is better.
> 
> But in this case `PaletteType` is practically an `int` (or something similar I suppose). I don't know how it works underneath in rust but if copying a `PaletteType` means copying an int and passing it by ref means passing a pointer, then passing it by value would be faster (you don't have the dereferencing part and since it's an `int` it will probably be passed via registers).
> 
> EDIT: I checked with godbolt and it seems that in rust the code for the two ways are basically the same (+ a copy in the `PaletteType` case) (with opt-level=0) because even when you pass by value it still dereferences since the Enum is stored in the heap in any case.
> 
> Also, in the case of optimizations they compile to the same assembly.
> 
> So I don't know, I would go for a `PaletteType` since it's "cleaner" but boh, no strong opinions since it's basically the same.
> 

Totally agree with you that in this case there is no difference. At first I coded it with `PaletteType` for the same reason. @guerinoni ask me to change it. Maybe because it is cleaner (I don't know how much) the sign of the method, but it is less cleaner the use of it as you need to add `.clone()` at the end :) 
Personally I prefer always reference if it possible, because you don't change the ownership of the variable, you don't need to copy it and so you can use it more times.

AlessioC31 commented:
> > > > Why do we pass &PaletteType instead of PaletteType in some functions?
> > > 
> > > 
> > > To avoid `clone()` when you pass it many times in other functions. In other case, without clone the variable exits from the scope and it will be lost. With `&PaletteType` we don't copy uselessly the value, but we are just passing its reference. For this specific case the difference is not so big because we don't need to pass by value a big amount of data, but anyway I think, when it is possible, this approach is better.
> > 
> > 
> > But in this case `PaletteType` is practically an `int` (or something similar I suppose). I don't know how it works underneath in rust but if copying a `PaletteType` means copying an int and passing it by ref means passing a pointer, then passing it by value would be faster (you don't have the dereferencing part and since it's an `int` it will probably be passed via registers).
> > EDIT: I checked with godbolt and it seems that in rust the code for the two ways are basically the same (+ a copy in the `PaletteType` case) (with opt-level=0) because even when you pass by value it still dereferences since the Enum is stored in the heap in any case.
> > Also, in the case of optimizations they compile to the same assembly.
> > So I don't know, I would go for a `PaletteType` since it's "cleaner" but boh, no strong opinions since it's basically the same.
> 
> Totally agree with you that in this case there is no difference. At first I coded it with `PaletteType` for the same reason. @guerinoni ask me to change it. Maybe because it is cleaner (I don't know how much) the sign of the method, but it is less cleaner the use of it as you need to add `.clone()` at the end :) Personally I prefer always reference if it possible, because you don't change the ownership of the variable, you don't need to copy it and so you can use it more times.

But if we derive `Copy` you don't need to use `.clone()` everytime, right? I ask because I genuinely don't know

guerinoni commented on emu/src/render/palette_color.rs:
```diff
@@ -0,0 +1,143 @@
+
+/// Memory info about BG palette
+pub const BG_PALETTE_ADDRESS: u32 = 0x05000000;
+pub const _BG_PALETTE_SIZE: usize = 0x200;
```
Not used?

gallottino replied:
Yes, I need it for later.

guerinoni commented on emu/src/render/palette_color.rs:
```diff
@@ -0,0 +1,143 @@
+
+/// Memory info about OBJ palette
+pub const OBJ_PALETTE_ADDRESS: u32 = 0x05000200;
+pub const _OBJ_PALETTE_SIZE: usize = 0x200;
```
Also this, is it not used?

gallottino replied:
Same of above.

guerinoni reviewed (approved):
Don't know if we miss some tests for `Bits` trait on all cases (u8, 816, u32, u128)
